sample-prep/Biomek_pooling_VI.py

```diff
@@ -29,107 +43,287 @@ def main(process_id, file_id):
+    
+    for input in sorted(inputs, key=sort_key):
+        source_location.append(input.location)
+        sample_name.append(input.name)
+        molarity.append(input.udf.get('Molarity TS (nM)'))
+        project_account.append(input.udf.get('Project Account'))
+        genome_size.append(input.udf.get('Genome Size (Mbp)'))
+        species.append(input.udf.get('Species'))
+        input_container_type.append(input.container.type.name)
+        input_container_name.append(input.container.name)
+        sample_volumes.append(round(molMbp*input.udf.get('Genome Size (Mbp)')/input.udf.get('Molarity TS (nM)'),2))
 
-    #exit script if entered step with more than one pool
+    # loop for 'Enforce Min Volume Required' button on the step. 
+    # If true, find the minimum value, find its index.
+    # Recalculate molMbp to achieve min_vol_required. New value put to the process udf for info to the user. volume_threshold_plotcolor set to yellow
+    # finally a for-loop to update sample_volumes using the new fmol/Mbp
 
-    unique_pools_dict = dict(zip(list(Pool_names),[list(Pool_names).count(i) for i in list(Pool_names)]))
-    Pool_names_unique = (list(unique_pools_dict.keys()))
+    if enforce_min_volume == True:
+        min_volume_in_list = min(sample_volumes)
+        min_volume_in_list_index = sample_volumes.index(min_volume_in_list)
+        if min_volume_in_list < min_vol_required: 
+            molMbp = (min_vol_required*molarity[min_volume_in_list_index])/genome_size[min_volume_in_list_index]
+            process.udf['fmol/Mbp'] = molMbp
+            process.put()
 
-    if len(Pool_names_unique) > 1:
-        print("This step only supports one pool at the time. Return to ice bucket and repeat with a single pool")
-        sys.exit(1)
+            index = 0
+            for input in sorted(inputs, key=sort_key):
+                sample_volumes[index] = (round(molMbp*input.udf.get('Genome Size (Mbp)')/input.udf.get('Molarity TS (nM)'),2))
+                index += 1
+
+    # volume_threshold_plotcolor to allow conditional coloring on plot
+    # adjusting volume to max_vol_available if enforce_max_vol
+    # over or under threshold = red, within threshold = palegreen, adjusted volume to max = yellow 
+
+    volume_threshold_plotcolor = []
+
+    for volumes in sample_volumes:
+        if volumes > max_vol_available or volumes < min_vol_required:
+            volume_threshold_plotcolor.append("red")
+        else:
+            volume_threshold_plotcolor.append("palegreen") 
+
+    if enforce_max_vol == True:
+        for i in range(len(sample_volumes)):
+            if sample_volumes[i] > max_vol_available:
+                sample_volumes[i] = max_vol_available
+                volume_threshold_plotcolor[i] = "yellow"
 
-    #write csv-file
+    if enforce_min_volume == True:
+        if min_volume_in_list < min_vol_required:
+            volume_threshold_plotcolor[min_volume_in_list_index] = "yellow"
+
```

Comment:
On a side note, this plot is just cool!

Reply:
Even cooler in the new version. Now placed in a neat loop in a separate function. Now I am also plotting a second line with percentage reduced for any samples with low concentration that have been forced to the max available volume (used some time to find a way to shift the second line on the discrete y axis). And I have added a new color to separate sample used to recalculate fmol/Mbp and those that have been forced down in volume.